### PR TITLE
Code changes

### PR DESCRIPTION
**Pull request by skuzzymiglet**
+ Channel recieves using range
+ No == with nil (which is never true)
+ Better error handling

### CODE REVIEW

**Comment by skuzzymiglet** [comment]
@Stargarth I'm working on the TOML config now, but I realized you're using a `map[string]interface{}` for blocks rather than a struct. I'll have to refactor that!

**Comment by Ldaxar** [comment]
Hey @skuzzymiglet !

Thanks for contributions. I'm currently on vacation and I'm too lazy to thoroughly review and test these changes. I should be able to do this in around 2 weeks. After quick glance, most of them seem ok. Change with timer.Tick is particularly nice. 

Just one thing, it's much better to make small changes. If there is a bunch of changes in pull request and just one of them is fishy, then maintainer has to reject all of them. This is frustrating for everybody so it's better to avoid it :P .

**Comment by skuzzymiglet** [comment]
Don't bother with the commits. I'd look at the changes as a whole. No
functionality has been added. For example, when
updating the bar fails, it exits (rather than doing an infinite loop).
If something bothers you, I can revert it. There's not much big change
really.

**Comment by Ldaxar** [comment]
Hi @skuzzymiglet !

I have to reject this pull request because of functional changes. In few places you alter the flow of the program by getting rid of if-else structs. I noticed this pattern in your changes:

1. Do something that is error prone
2. Check for error
3. If error exists log the error
4. Progress further with execution

This is a straight way to get strange crashes and unintended behavior. If error is detected the error message should be thrown and program should halt/recover. 

**For example:**
1. Do something that is error prone
2. Check for error
3a. If error exists log the error, try to recover or terminate
3b. If error is NULL progress further with execution

Some might say that it is nitpicking as correct error message is thrown before program continues to do whatever it needs to do in that illegal state(which will probably result in crash). But I believe that correct program should never crash beside situations that you obviously can't account for (hardware failures, explosion on sun surface which create massive electromagnetic waves that can flip few bits in your memory etc.).

Nonetheless thanks for these changes as I have already shamelessly stolen few of your ideas (ticker and smarter config directory search). Should you want to add rest of your changes, just create new pull request that complies with remarks made above. Also please try to make pull requests that are more focused on single thing (renaming, formatting, prints to logging etc.). 2/3 of your changes could have been merged right away 😄  

Thanks for contributing to goblocks.

File: util/utils.go [hunk 1]
```diff
@@ -20,13 +20,11 @@ type Change struct {
...
 	u, err := time.ParseDuration(duration)
-	if err == nil {
-		for {
-			send <- true
-			time.Sleep(u)
-		}
-	} else {
-		fmt.Println("Couldn't set a scheduler due to improper time format: " + duration)
+	if err != nil {
+		log.Fatalf("Couldn't set a scheduler due to improper time format: %s\n", duration)
```
Comment (skuzzymiglet) [inline util/utils.go:24]: You know `log.Fatalf` stops the program right?

**Review by skuzzymiglet — COMMENTED** [review]
It's idiomatic Go to check for the error and exit/return/log, and carry on the flow on the left - see: https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88

**Comment by skuzzymiglet** [comment]
Additionaly,  I'm gonna revert commits/mess around with this pull request rather than starting anew (`git revert`)

File: util/utils.go (hunk 1 above)
Comment (Ldaxar) [reply util/utils.go:24]: Well, I actually didn't know it ! Despite the fact that this software is written in Go, I don't write that much Go 😸. You can disregard my previous post. Sorry for not checking it out, it just didn't come to my mind that logging statement could terminate program as well!

**Comment by skuzzymiglet** [comment]
Seems you've recreated most of my commits on master. You're welcome to recreate the error handling.

File: goblocks.go [hunk 2]
```diff
@@ -5,91 +5,96 @@ import (
...
+	"path/filepath"
 	"strings"
 )
 
 type configStruct struct {
-	Separator string
+	Separator          string
 	ConfigReloadSignal int //currently unused
-	Actions []map[string]interface{}
+	Actions            []map[string]interface{}
```
Comment (skuzzymiglet) [inline goblocks.go:18]: Why are you using a type-unsafe `map[string]interface{}` here? Can't you use a struct?

File: util/utils.go [hunk 3]
```diff
@@ -1,49 +1,49 @@
...
+	}
+	for range time.Tick(u) {
+		send <- true
 	}
 }
+
 //Run any arbitrary POSIX shell command
-func RunCmd(blockId int, send chan Change, rec chan bool, action map[string]interface{} ) {
+func RunCmd(blockID int, send chan Change, rec chan bool, action map[string]interface{}) {
 	cmdStr := action["command"].(string)
```
Comment (skuzzymiglet) [inline util/utils.go:33]: This for example. What if command is not a key? And you have to extract the underlying type (which is ugly). Precisely why we have structs

File: goblocks.go (hunk 2 above)
Comment (Ldaxar) [reply goblocks.go:18]: Not necessarily. The thing is that we have no idea what action block will contain, they require only two parameters - "command" and "timer".  This ties back to "builtIn" functionality that receives raw action block and extract whatever it needs to configure itself. For example "format" parameter is irrelevant for "sh" blocks but it is relevant for CPU and Memory block. In theory you could add waaaay more to action blocks and have corresponding built in functionality that makes use of it. Actually, you can extend like that without much of in-code changes. Just put whatever you need in action JSON block and extract it in your builtIn code (through action["whatever"]). As far as I remember this is exactly how "format" parameter was introduced. Somebody added 1 line in goblocks.json, and 1 line in CPU and memory blocks that used it to format themselves. 

I guess you could make a struct like:
```
type actionStruct struct {
	Command string
        Timer string
	Rest []map[string]interface{}
}
```
But that doesn't get rid of []map[string]interface{}. Moreover, you are introducing extra code to essentially get rid of 2 casts(you still need to make an awkward check for prefix/suffix as they are also optional). However, there might be some Go specific technique that I'm not aware of 😅.  

tl;dr: There is no struct because action blocks actually do not have a structure. You can think of goblocks.json as a "collection of config files" where each action is a "config file" for a specific block.

File: util/utils.go (hunk 3 above)
Comment (Ldaxar) [reply util/utils.go:33]: Yeah I guess I should check for "command" and "timer" parameters. As to the awkward handling of actions I explained it in previous review.

File: goblocks.go (hunk 2 above)
Comment (skuzzymiglet) [reply goblocks.go:18]: There is: `nil`.  If a value's not there, it'll be nil. Put everything into the struct, use it as needed. If you want safety, check `value != nil` before proceeding. Although your different block types _shouldn't_ be accessing fields they know will be nil for them.

On getting rid of two casts, it's worth it. Type safety is one of the reasons to use Go - bugs are reported early. What if the type you plonk in becomes a `float64` (because of a bug), and you  try to find the underlying `string` - boom!  A `panic`!

Comment (Ldaxar) [reply goblocks.go:18]: > Put everything into the struct, use it as needed.
>
I didn't get it. Could you provide me with an example of struct that has variable number of attributes?

Comment (skuzzymiglet) [reply goblocks.go:18]:
```
type Action struct {
	Command        string
	UpdateSignal   int
	Timer          string
	Prefix, Suffix string
	Format         string
}

```

json.Unmarshal will leave unassigned ones at nil

Comment (skuzzymiglet) [reply goblocks.go:18]: Side note: could you make updateSignal optional? Sucks to have to put it in each entry.

Side note 2: I removed the builtins in my fork. I prefer to use the commands if they're available. Consistency

Comment (Ldaxar) [reply goblocks.go:18]: > ```
> type Action struct {
> 	Command        string
> 	UpdateSignal   int
> 	Timer          string
> 	Prefix, Suffix string
> 	Format         string
> }
> ```
> 
> json.Unmarshal will leave unassigned ones at nil

I think you missed the point. This does not address the flexibility that I was talking about.

> Side note: could you make updateSignal optional? Sucks to have to put it in each entry.
>
Yeah, that could be done. Create an issue with "enhancement" tag. I tend to forget things.

Comment (skuzzymiglet) [reply goblocks.go:18]: Flexibility? You don't want flexibility, you want type-safety. (this isn't python 😉). If they're not configured, nil, there you go. What else do you want?

Comment (Ldaxar) [reply goblocks.go:18]: > Flexibility? You don't want flexibility, you want type-safety. (this isn't python 😉).
>
I find this statement really silly. Of course I want flexibility. Besides that, since when programming language dictates architecture? It's a property that stems from design. You can have flexible software in any language Go/COBOL/Python/C etc.
> What else do you want?
>
I've said this before so let me reiterate. I want to be able to access **variable** number of attributes. 

You added "Format" because you know right now that format is used. So if I add a new weather block (wttr.in kinda sucks), should I add properties such as location/temperature scale/etc. to action struct? Okay and then somebody adds a block that reads S&P 500 prices, I guess him/her should add his/her properties to action block too. You want a block that shows current number of COVID-19 cases in your country? Add all your properties to Action struct. Then we end with Action struct that potentially has 100 attributes and for every block it tries to map 95% properties that do not exist. If that is not bloat than I don't know what is. 

With current set-up new builtIn programs have literally 0 effect on rest of goblocks.

Comment (skuzzymiglet) [reply goblocks.go:18]: > bloat

Well, `nil` takes up no memory. And if you think adding those fields to the struct is bloat, just don't add the functionality. Adding them to a map would be "bloat" - just uncharted bloat. Do you really want to `ok`-check everything, and cast the types? That's gonna be buggy.

Also, here's an idea - compose the structs:

```
type Block struct {
timer string
prefix, suffix string
}

type CommandBlock struct {
Block
command string
}


type WeatherBlock {
Block
location string
format string
}
```

(Adding those is bloat imo, but your call)

Comment (Ldaxar) [reply goblocks.go:18]: > Well, nil takes up no memory.
>
I didn't research it but based on how it works in other programming languages I'm 99% sure that this statement is just false.

Even when you have a nil value you still need a reference (that points to nothing) but that reference is still a pointer that has a size that is equal to a word length of particular computer architecture that you are working on.

> Adding them to a map would be "bloat" - just uncharted bloat.
> 

Let's assume that you have a struct as following:
```
type actionStruct struct {
	Command string
        Timer string
        Sth1 string
        Sth2 string
}
```
There are 10 blocks and Sth1, Sth2 are relevant for only one respective block. In that scenario we have 10 copies of Sth1 and 10 copies of Sth2. In total we have 20 references where 18 of them point to nothing and 2 of them are pointing to something relevant. On 64 bit system that would be 144B of wasted space.

This does not occur with a map. Action blocks that do not have Sth1 and Sth2 just don't have those entries. 0B wasted space.

It is also worth to note that Go JSON parser will look for Sth1, Sth2 and that's just a waste of computation, because we know beforehand that they do not occur in 90% of entries.

>  Do you really want to ok-check everything, and cast the types? That's gonna be buggy.
> 

Interpretation of interface map is responsibility of builtIn programs not goblocks. Goblocks just takes whatever was in JSON file, gets minimal subset that it needs ( command etc.) and ships rest to appropriate program. As to ok checking command and prefix/suffix, that will be changed.

Comment (skuzzymiglet) [reply goblocks.go:18]: That's why you stick to explicit, fixed data structures. This is hardly a problem for my fork, as I am staying with oh-so-flexible commands rather than bolting on loads of features. I honestly don't see the need for the builtins, but if you do desire them keep some part in a struct and add a dictionary. It'll be much easier to understand and read.

PS: 144B of wasted space? We're programmers, we make things, we read, we don't obesses about tiny bits of memory 😉. That's why you're not writing this program in assembly!

**Comment by skuzzymiglet** [comment]
Are you willing to merge some changes or do you want me to close this PR? Seems you've copied most of the work under your name on master

**Comment by Ldaxar** [comment]
I can merge it if you will make changes that resolve conflict. Sorry for the trouble - it's my fault that I didn't check that log.fatal terminates the program and as a result I made invalid assumptions. 

As to copying your changes - yeah, I did. If you want me to give you some recognition in README just say a word.

**Comment by skuzzymiglet** [comment]
I don't think there is a point merging this. If you take a look at "Files Changed", you can just pick changes from there and put them into master. I don't think you merged the error handling, which needs to be done imo as you are moving code rightwards in an `err == nil` statement

**Comment by Ldaxar** [comment]
I've added enhanced error handling so basically all changes from this pull request have been added. As a result I'm closing it.
